Avoid heap allocation in jsEventTargetCast()

cdumez opened the pull request:
#### fd7c1f2261ac273ed5b757ea85426aef4d00f37c
<pre>
Avoid heap allocation in jsEventTargetCast()
<a href="https://bugs.webkit.org/show_bug.cgi?id=254672">https://bugs.webkit.org/show_bug.cgi?id=254672</a>

Reviewed by Darin Adler.

* Source/WebCore/bindings/js/JSEventTargetCustom.cpp:
(WebCore::jsEventTargetCast):
* Source/WebCore/bindings/js/JSEventTargetCustom.h:
(WebCore::JSEventTargetWrapper::JSEventTargetWrapper):
(WebCore::JSEventTargetWrapper::isValid const):
(WebCore::JSEventTargetWrapper::wrapped):
(WebCore::JSEventTargetWrapper::operator JSC::JSObject&amp;):
(WebCore::IDLOperation&lt;JSEventTarget&gt;::call):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::initMouseEventQuirk):

Canonical link: <a href="https://commits.webkit.org/262294@main">https://commits.webkit.org/262294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/077a2af025f98a76c67a313edf8e0ec1d304bbba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1130 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1808 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/996 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1114 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1241 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1188 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1138 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1045 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1701 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1093 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1043 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1034 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1031 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1072 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2126 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1077 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1017 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1052 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/285 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1098 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/js/JSEventTargetCustom.cpp:
```diff
@@ -62,15 +62,15 @@ EventTarget* JSEventTarget::toWrapped(VM&, JSValue value)
 {
     if (auto* target = jsDynamicCast<JSEventTarget*>(thisValue))
-        return makeUnique<JSEventTargetWrapper>(target->wrapped(), *target);
+        return JSEventTargetWrapper { target->wrapped(), *target };
```
I think we don’t need the type name in these `return` statements since this isn't using `optional`.

darinadler commented on Source/WebCore/bindings/js/JSEventTargetCustom.h:
```diff
@@ -34,23 +34,25 @@ namespace WebCore {
     { }
 
-    EventTarget& wrapped() { return m_wrapped; }
+    bool isValid() const { return m_wrapped; }
```
We have competing idioms for this in WebKit. Sometimes `isValid`, sometimes `isNull`, in the standard library `has_value`, sometimes `explicit operator bool` and sometimes `operator!`. I wish it was more consistent.

cdumez committed:
Avoid heap allocation in jsEventTargetCast()
https://bugs.webkit.org/show_bug.cgi?id=254672

Reviewed by Darin Adler.

* Source/WebCore/bindings/js/JSEventTargetCustom.cpp:
(WebCore::jsEventTargetCast):
* Source/WebCore/bindings/js/JSEventTargetCustom.h:
(WebCore::JSEventTargetWrapper::JSEventTargetWrapper):
(WebCore::JSEventTargetWrapper::isValid const):
(WebCore::JSEventTargetWrapper::wrapped):
(WebCore::JSEventTargetWrapper::operator JSC::JSObject&):
(WebCore::IDLOperation<JSEventTarget>::call):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::initMouseEventQuirk):

Canonical link: https://commits.webkit.org/262294@main
